diff --git a/src/router_core/connections.c b/src/router_core/connections.c
index 2abb7fc..381f231 100644
--- a/src/router_core/connections.c
+++ b/src/router_core/connections.c
@@ -89,6 +89,7 @@ qdr_connection_t *qdr_connection_opened(qdr_core_t            *core,
     conn->connection_info->role = conn->role;
     conn->work_lock = sys_mutex();
 
+    qd_log(core->log, QD_LOG_INFO, "Connection opened (incoming=%s, container_id=%s, host=%s)", incoming ? "true" : "false", remote_container_id, connection_info->host);
     if (vhost) {
         conn->tenant_space_len = strlen(vhost) + 1;
         conn->tenant_space = (char*) malloc(conn->tenant_space_len + 1);
@@ -108,6 +109,7 @@ qdr_connection_t *qdr_connection_opened(qdr_core_t            *core,
 void qdr_connection_closed(qdr_connection_t *conn)
 {
     qdr_action_t *action = qdr_action(qdr_connection_closed_CT, "connection_closed");
+    qd_log(conn->core->log, QD_LOG_INFO, "Connection closed (incoming=%s, container_id=%s, host=%s)", conn->incoming ? "true" : "false", conn->connection_info->container, conn->connection_info->host);
     action->args.connection.conn = conn;
     qdr_action_enqueue(conn->core, action);
 }
