diff --git a/src/router_node.c b/src/router_node.c
index abf80f4..64fe2fe 100644
--- a/src/router_node.c
+++ b/src/router_node.c
@@ -142,6 +142,7 @@ static qd_iterator_t *router_annotate_message(qd_router_t       *router,
             // the message has already passed through.
             //
             *link_exclusions = qd_tracemask_create(router->tracemask, trace);
+            qd_bitmask_set_all(*link_exclusions);

             //
             // Append this router's ID to the trace. 
